File: tests/src/main/java/com/github/klee0kai/test/lifecycle/di/PComponent.java
package com.github.klee0kai.test.lifecycle.di;

import com.github.klee0kai.stone.annotations.component.Component;
import com.github.klee0kai.stone.types.lifecycle.IStoneLifeCycleOwner;
import com.github.klee0kai.test.lifecycle.GoodPhone;
import com.github.klee0kai.test.lifecycle.OnePhone;
import com.github.klee0kai.test.lifecycle.di.qualifiers.DataStorageSize;
import com.github.klee0kai.test.lifecycle.di.qualifiers.RamSize;

@Component(
        qualifiers = {DataStorageSize.class, RamSize.class}
)
public interface PComponent {

    Structure structure();

    void inject(OnePhone onePhone);

    void inject(GoodPhone goodPhone, DataStorageSize dataStorageSize, RamSize ramSize);

    void inject(GoodPhone goodPhone, IStoneLifeCycleOwner lifeCycleOwner, DataStorageSize dataStorageSize, RamSize ramSize);
}

File: test_kotlin/src/main/kotlin/com/github/klee0kai/test/lifecycle/di/PComponent.kt
package com.github.klee0kai.test.lifecycle.di

import com.github.klee0kai.stone.annotations.component.Component
import com.github.klee0kai.stone.types.lifecycle.IStoneLifeCycleOwner
import com.github.klee0kai.test.lifecycle.GoodPhone
import com.github.klee0kai.test.lifecycle.OnePhone
import com.github.klee0kai.test.lifecycle.di.qualifier.DataStorageSize
import com.github.klee0kai.test.lifecycle.di.qualifier.RamSize

@Component(qualifiers = [DataStorageSize::class, RamSize::class])
interface PComponent {
    fun structure(): Structure?
    fun inject(onePhone: OnePhone?)
    fun inject(goodPhone: GoodPhone?, dataStorageSize: DataStorageSize?, ramSize: RamSize?)
    fun inject(
        goodPhone: GoodPhone?,
        lifeCycleOwner: IStoneLifeCycleOwner?,
        dataStorageSize: DataStorageSize?,
        ramSize: RamSize?,
    )
}